Replace turbo-combine-reducers with combineReducers from Redux

jsnajdr opened the pull request:
This is still an experiment, I'm opening a PR mainly to get performance test results.

There are two major problems with `combineReducers` in Redux.

First, it doesn't allow slice reducers to ever return `undefined` state. `undefined` means uninitialized and every reducer is supposed to return an "initialized" inital state when `undefined` state is passed. We have many reducers that use `undefined`, and this PR has to modify them to use `null` instead. Otherwise, `combineReducers` will throw an exception and the app will crash. The prohibition on `undefined` will be a breaking change that we probably can't afford to do.

Second, it logs a warning when the combined reducer function is called with a state that contains unknown keys, with no reducer for them. For example, consider combining `foo` and `bar` reducers with `combineReducers( { foo, bar } )`. Then, when you pass an object with shape `{ foo, bar, baz }` as `state`, `combineReducers` doesn't like the `baz` field, because there is no sub-reducer to pass it to. The warning is:
```
Unexpected key "baz" found in preloadedState argument passed to createStore.
Expected to find one of the known reducer keys instead: "foo", "bar".
Unexpected keys will be ignored.
```

The `core/blocks` store very often uses an enhancer pattern that leads to these "unexpected key" warnings:
```js
function withIgnoredBlockChange( reducer ) {
  return ( state, action ) => {
    const nextState = reducer( state, action );
    if ( nextState !== state ) {
      nextState.isIgnoredChange = action.type === 'RECEIVE_BLOCKS';
    }
    return nextState;
  };
}
```
This enhancer calls the inner reducer and then adds the `isIgnoredChange` field to the existing ones. If the inner reducer is a result of `combineReducers`, it will complain about the unknown key. Although everything works, there's nothing really wrong with this pattern.

For these two reasons, we'll probably have to write our own version of `combineReducers`.

Fixes #42513.

jsnajdr committed:
Replace turbo-combine-reducers with combineReducers from Redux

tyxla commented on packages/edit-site/src/components/editor-canvas-container/index.js:
```diff
@@ -90,7 +90,7 @@ function EditorCanvasContainer( {
 		}
 		setIsListViewOpened( showListViewByDefault );
-		setEditorCanvasContainerView( undefined );
+		setEditorCanvasContainerView( null );
```
Note that changing all reducers to use `undefined` instead of a `null` is likely to end up being a regression, since some of the reducers might make a strict differentiation between the `null` and `undefined`. This is especially valid for custom `@wordpress/data` stores that are already using `undefined` as a valid value somewhere in their reducer trees.

With that in mind, I think we might have to consider adding our own wrapper around the Redux `combineReducers` that handles `undefined` in a way that allows supporting it as a valid value.

tyxla replied:
@jsnajdr what are your thoughts on the above comment?

jsnajdr replied:
Now, after I abandoned the idea to use `combineReducers` from Redux and created my own version instead, `undefined` state is no longer a concern. And neither are extra keys in state. `@wordpress/data` reducers can continue to use them without any change.

However, this PR still contains changes to Gutenberg's data stores that remove `undefined` state and replace it with `null`. They're no longer strictly needed, but still seem to me like a good code quality improvement.

But these changes are not worth doing if they introduce any backward compatibility issues for the existing Gutenberg stores. But are there any? And if there are, can they be solved?

For example, the `setEditorCanvasContainerView( undefined )` call no longer works and has to be changed to `setEditorCanvasContainerView( null )`. But that can be easily avoided by defaulting the action parameter to `null`: `( view = null ) => ...`. I haven't 100% verified this yet, I only know that unit and e2e tests are green.

jsnajdr committed:
Fix createReduxStore type generation

jsnajdr commented:
Here are performance tests results. I don't see any significant regressions. Some tasks are faster, some are slower, and my intuitive guess is that none of them is outside the random fluctuation threshold.

```
post-editor
┌──────────────────────┬──────────────────────────────────────────┬───────────────┐
│       (index)        │ 6a0c6c625075126721a3e22cbb79fad3b54755fd │     trunk     │
├──────────────────────┼──────────────────────────────────────────┼───────────────┤
│    serverResponse    │               '337.35 ms'                │  '329.4 ms'   │
│      firstPaint      │                '610.9 ms'                │  '590.65 ms'  │
│   domContentLoaded   │               '1295.25 ms'               │ '1278.75 ms'  │
│        loaded        │               '1296.45 ms'               │  '1279.8 ms'  │
│ firstContentfulPaint │               '2296.75 ms'               │ '2309.25 ms'  │
│      firstBlock      │              '15015.75 ms'               │ '15173.65 ms' │
│         type         │                '67.38 ms'                │  '72.98 ms'   │
│       minType        │                '60.84 ms'                │  '68.72 ms'   │
│       maxType        │                '78.58 ms'                │  '92.13 ms'   │
│    typeContainer     │                '22.43 ms'                │  '27.19 ms'   │
│   minTypeContainer   │                '19.16 ms'                │  '20.36 ms'   │
│   maxTypeContainer   │                '27.54 ms'                │  '35.26 ms'   │
│        focus         │               '124.28 ms'                │  '96.36 ms'   │
│       minFocus       │               '112.19 ms'                │  '84.35 ms'   │
│       maxFocus       │               '141.69 ms'                │  '130.34 ms'  │
│     inserterOpen     │                '40.31 ms'                │  '47.15 ms'   │
│   minInserterOpen    │                '37.38 ms'                │  '35.43 ms'   │
│   maxInserterOpen    │                '44.13 ms'                │  '122.86 ms'  │
│    inserterSearch    │                 '9.1 ms'                 │   '9.55 ms'   │
│  minInserterSearch   │                '8.15 ms'                 │   '8.3 ms'    │
│  maxInserterSearch   │                '10.36 ms'                │  '12.17 ms'   │
│    inserterHover     │                '56.86 ms'                │  '53.67 ms'   │
│   minInserterHover   │                '50.58 ms'                │  '48.04 ms'   │
│   maxInserterHover   │                '62.61 ms'                │  '60.66 ms'   │
│     listViewOpen     │               '214.57 ms'                │  '214.13 ms'  │
│   minListViewOpen    │               '200.63 ms'                │  '193.45 ms'  │
│   maxListViewOpen    │               '235.62 ms'                │  '235.6 ms'   │
└──────────────────────┴──────────────────────────────────────────┴───────────────┘

site-editor
┌──────────────────────┬──────────────────────────────────────────┬───────────────┐
│       (index)        │ 6a0c6c625075126721a3e22cbb79fad3b54755fd │     trunk     │
├──────────────────────┼──────────────────────────────────────────┼───────────────┤
│    serverResponse    │                '141.2 ms'                │  '161.15 ms'  │
│      firstPaint      │                '650.7 ms'                │  '596.25 ms'  │
│   domContentLoaded   │               '1330.75 ms'               │  '1306.3 ms'  │
│        loaded        │               '1331.6 ms'                │ '1307.05 ms'  │
│ firstContentfulPaint │               '1555.45 ms'               │  '1529.5 ms'  │
│      firstBlock      │               '13297.2 ms'               │ '13061.65 ms' │
│         type         │                '77.11 ms'                │  '80.73 ms'   │
│       minType        │                '69.41 ms'                │  '71.48 ms'   │
│       maxType        │                '97.75 ms'                │  '105.48 ms'  │
└──────────────────────┴──────────────────────────────────────────┴───────────────┘
```

jsnajdr committed:
Switch to custom combineReducers that doesn't do any checks

jsnajdr committed:
Better version of combineReducers

jsnajdr committed:
getCurrentTheme: don't call getEntityRecord with null key

tyxla reviewed (commented):
This is great work, thanks for working on it @jsnajdr!

As we discussed separately, it might be a good idea to move the `undefined` -> `null` changes to a separate PR, since they might be controversial and potentially breaking for some use cases. 

I also think we can add some unit tests to cover our custom `combineReducers` implementation.

Once those two points are addressed, I think we should be good to go.

jsnajdr committed:
Unit tests for combineReducers

jsnajdr commented on packages/core-data/src/selectors.ts:
```diff
@@ -936,6 +936,9 @@ export function hasRedo( state: State ): boolean {
 export function getCurrentTheme( state: State ): any {
+	if ( ! state.currentTheme ) {
+		return null;
+	}
```
This is a fix that was forced by the "`null` state instead of `undefined`" changes, and it can be merged even though the `state.currentTheme` remains `undefined`. `getEntityRecord` should never get a fourth argument (`key`) that's not a valid string, its type doesn't allow that. The `State` type for the `getCurrentTheme`'s `state` argument "lies" about the type of `state.currentTheme`. It can be `undefined` (or `null`) when there is no active theme or when the active theme is not yet fetched from the REST endpoint.

tyxla replied:
I realize this is pre-existing, but there are places in the codebase that rely on the `getCurrentTheme()` to always return a non-nullish value, for example:

https://github.com/WordPress/gutenberg/blob/dd7e10adb416efeb213b79ffb5a97a23e454fc1e/packages/block-library/src/pattern/edit.js#L23

Should we be concerned that those potentially might cause issues under some conditions?

jsnajdr replied:
If `state.currentTheme` is not a valid theme slug, then `getEntityRecord` would return `undefined` anyway, because the state never has a theme whose slug (ID) would be `undefined` or `null`. So, it's not that we're now returning `null` where a theme object would be previously returned.

The existing places in the codebase are incorrect, because `getCurrentTheme()` is not guaranteed to return a valid theme. The code relies on the fact that by the time it runs, the `/wp/v2/themes?status=active` REST request has already finished loading.

tyxla replied:
Agreed - the existing code would throw an error before with the `undefined`, and will throw an error now with the `null`. 

IMHO these cases can be handled in another PR, and are something that an eventual TS migration will catch anyway.

jsnajdr commented on packages/block-library/src/embed/test/index.native.js:
```diff
@@ -138,7 +138,12 @@ const mockEmbedResponses = ( mockedResponses ) => {
+				stylesheet: 'test-theme',
+				theme_supports: { 'responsive-embeds': true },
+			},
+		];
```
Related to the `state.currentTheme` fix. The `stylesheet` field contains the ID of the theme, and if the test mock doesn't supply it, the `state.currentTheme` becomes `undefined`.

jsnajdr commented:
> Once those two points are addressed, I think we should be good to go.

Both points are addressed now: I removed the changes for `undefined` state values, and added unit tests for `combineReducers`.

tyxla commented on packages/data/src/redux-store/index.js:
```diff
@@ -23,6 +23,8 @@ import metadataReducer from './metadata/reducer';
 import * as metadataSelectors from './metadata/selectors';
 import * as metadataActions from './metadata/actions';
 
+export { combineReducers };
```
I think this needs a CHANGELOG entry.

jsnajdr replied:
Good point, added a changelog entry for the `data` package.

tyxla commented on packages/data/src/redux-store/combine-reducers.js:
```diff
@@ -0,0 +1,17 @@
+export function combineReducers( reducers ) {
```
This is more or less a simplified version of the Redux `combineReducers()`: https://github.com/reduxjs/redux/blob/957aed6e23624113b4d0ff6d9f277f25fca90b38/src/combineReducers.ts#L123

Should we add a mention in CREDITS.md?

jsnajdr replied:
Probably not, I think it's sufficiently different from Redux. There are only so many ways to write a `combineReducers` function, it's very hard to be completely original 🙂 Also, Gutenberg doesn't even have a `CREDITS.md` file.

tyxla replied:
Fair enough 👍

tyxla commented on packages/data/src/redux-store/test/combine-reducers.js:
```diff
@@ -0,0 +1,58 @@
+		} );
+	} );
+
+	it( 'does not mind unknown state keys', () => {
```
Redux's `combineReducers` will also complain if there are no reducer subkeys. We could add a test for that too, although it would be a rare scenario.

jsnajdr replied:
I don't understand, do you mean calling `combineReducers( {} )`? I think that's a valid case, it returns a valid "zero" reducer that always returns `{}`. Something like empty `{}` TypeScript type.

tyxla replied:
Yes, I was talking about `combineReducers( {} )`. AFAIK the Redux version won't allow this, while our version allows it. So that makes it a good case for a unit test IMHO.

tyxla commented on packages/data/src/redux-store/combine-reducers.js:
```diff
@@ -0,0 +1,17 @@
+export function combineReducers( reducers ) {
```
Should we add a failproof check to ensure `reducers` is a proper object? Should be cheap IMHO. If you agree we should, it could be yet another case to add a test for.

Alternatively, we could introduce `combineReducers` as a typed function. Then we wouldn't be worried about non-object `reducers` arguments.

jsnajdr replied:
Neither Redux nor `turbo-combined-reducers` do such a check on `reducers`. Both of them do an `Object.keys( reducers )` call at the very beginning, that's where they crash if it's not a proper object.

`@wordpress/data` re-exports the Redux type for `combineReducers`, so the function has been already typed, and continues to be.

tyxla replied:
Agreed, no need to overcomplicate 👍

tyxla reviewed (commented):
This looks great overall, thanks for working on it @jsnajdr!

I think it's quite close and I've left a few questions - let me know what you think.

jsnajdr committed:
Add changelog entry

tyxla reviewed (approved):
Looks and works great, thanks for addressing it @jsnajdr!

I think we could add a unit test for `combineReducers( {} )` before shipping, but that's not a blocker.

🚀

jsnajdr committed:
Add unit test for unit reducer

jsnajdr commented:
> I think we could add a unit test for `combineReducers( {} )` before shipping

It's added, now I'm going to 🚢

noahtallen commented:
After installing this update, I'm seeing a lot of errors like `Property 'analyzer' does not exist on type '{}'.` I think this comes from type definitions like this: `export type State = ReturnType< typeof reducers >;`. This currently turns into `type State: {}`, because the return type of `combineReducers` is now `{}`, even when the input reducers are fully typed.

tyxla commented:
Fix for the type problem is here: #55321
